timestamp-impl/src/test/java/com/palantir/timestamp/PersistentUpperLimitTest.java

```diff
@@ -111,4 +126,44 @@ public void shouldNotChangeTheCurrentUpperLimitIfItFailsToPersist() {
 
         assertThat(upperLimit.get()).isEqualTo(INITIAL_UPPER_LIMIT);
     }
+
+    @Test
+    public void increaseToAtLeastAcquiresLock() {
+        upperLimit.increaseToAtLeast(TIMESTAMP);
+        verify(clientLock, times(1)).get();
+    }
+
+    /**
+     * Force the first thread to run first and then have it wait to obtain the permit from the second thread. The
+     * second thread will not be able to release the permit, as the first thread is holding the lock.
+     */
+    @Test
+    public void increaseToAtLeastIsThreadSafe() throws InterruptedException {
+        Semaphore taskRunningSemaphore = new Semaphore(0);
+        Semaphore secondThreadSemaphore = new Semaphore(0);
+        doAnswer(invocation -> {
+                    taskRunningSemaphore.release();
+                    secondThreadSemaphore.tryAcquire(1L, TimeUnit.SECONDS);
+                    return null;
+                })
+                .when(boundStore)
+                .storeUpperLimit(anyLong());
+        TimestampBoundStore secondBoundStore = mock(TimestampBoundStore.class);
+        doAnswer(invocation -> {
+                    secondThreadSemaphore.release();
+                    return null;
+                })
+                .when(secondBoundStore)
+                .storeUpperLimit(anyLong());
+        PersistentUpperLimit identicalUpperLimit = new PersistentUpperLimit(secondBoundStore, clientLock);
+        Future<?> firstTask = taskExecutor.submit(() -> upperLimit.increaseToAtLeast(TIMESTAMP));
+        taskRunningSemaphore.tryAcquire(1, TimeUnit.SECONDS);
```

Comment:
maybe this is fine, but feels a bit suspicious that we're not considering the return value of `tryAcquire`?